### ISSUE TITLE
gr-soapy: support specifying arbitrary channels for SoapySDR stream

### ISSUE THREAD

**Issue by RomaVis** [gnuradio/gnuradio#6281]
### Feature Description

(TLDR? Skip to **TLDR** marker)

---

Hi,

I'm busy with implementing SoapySDR driver for a custom Rx device, which I'd like to use via gr-soapy later on. When I was working on implementation, I've stumbled upon something that looks like a limitation of gr-soapy.

To keep explanation concise, let's imagine a hypothetical situation - interfacing to a multi-channel I/Q digitizer board. There are 6 SMA jacks on the front panel, which correspond to three I/Q signal pairs. Each pair is routed to its own two-channel ADC. All ADCs are connected to an FPGA, which streams data over USB port to the PC. FPGA also implements control register (CSR) space, through which PC can control & monitor different aspects of HW operation. So in total there are three ADCs (0, 1 and 2), each of them providing its own I&Q data stream. ADCs are in many aspects independent from each other - e.g. they can be individually enabled/disabled, each ADC has its own sample clock, which is created by dividing down master FPGA clock, etc.

In terms of SoapySDR, it's clear that the whole board has to be managed by a single `SoapySDR::Device` instance (since there's a single USB interface to communicate with it, shared CSR space, etc.). Then, since most of controls in SoapySDR API are per-channel, it is reasonable to expose 3 Rx channels (0, 1 and 2) - which correspond to three ADCs. Each of them will at least have its own sample rate control. Now, since sample rates for channels may be different, and `SoapySDR::Stream` requires all participating channels to have the same sample rate, I've decided that the driver will support one `SoapySDR::Stream` per channel. To get data from channel 0, you call `Device::setupStream(channels=[0])` and you'll get a stream of CF32 samples from ADC 0. To get data from channel 2, call `Device::setupStream(channels=[2])` and you'll get CF32 samples from ADC 2. If you want to use all three ADCs at once, since channels can run at different sample rates, you create three separate single-channel streams and read them individually. So far so good.

When I've tried to bridge this to gr-soapy, things got tricky. Each gr-soapy block corresponds to a single `SoapySDR::Stream`. Thanks to automatic sharing of devices in SoapySDR device factory, I can create three `soapy.source` blocks which will refer to three different streams of the same device, which is exactly what I would like to do. However...

---
 **TLDR** 

... It turns out that gr-soapy does not support passing freestyle list of channels to `SoapySDR::Device::setupStream()` method. It is not possible to create a gr-soapy block which will call e.g. `setupStream(channels=[2])`. For some reason gr-soapy block API requires one to pass `nchan` (number of channels) instead of `channels` (list of channels), as can be seen from the signature of C++ constructor of `block_impl` class in gr-soapy:
```
block_impl::block_impl(int direction,
                       const std::string& device,
                       const std::string& type,
                       size_t nchan,
                       const std::string& dev_args,
                       const std::string& stream_args,
                       const std::vector<std::string>& tune_args,
                       const std::vector<std::string>& other_settings)
```
With this API, it always calls `setupStream()` with contiguous list of channels, starting from 0: it can do e.g. `channels=[0], [0,1], [0,1,2,3]..`, but it can not do `channels=[2,3]`.

Unless I'm missing something, this seems like an obvious limitation when working with multi-channel SoapySDR devices.

I see several ways how this can be implemented:
1. Modify gr-soapy block constructor signature
a. Replace `nchan: int` with `channels: List[int]`. This is the easiest option, but it will not be backwards-compatible with existing hand-written python/C++ flowgraphs that instantiate gr-soapy blocks, but it can be made compatible with existing GRC flowgraphs.
b. Replace `nchan: int` with `channels: Union[int, List[int]]` - iow, support passing both `int` and `List[int]` there. This will be backwards-compatible with existing code, but I'm not sure how easy is it to implement this in C++.
2. Keep constructor unchanged, but provide a new block method e.g. `set_channels(channels: List[int])`. The only good point about this is that existing block API stays unmodified - but otherwise this idea is IMO unreasonable. List of channels is not a runtime-modifiable parameter, and by all means belongs to block's constructor. Trying to do otherwise may make implementation very questionable (primarily because stream has to be completely re-created when list of channels is modified).

I can find some time to work on implementation once we agree on design.

### Feature Urgency

low (just an idea)

**Comment by willcode** [comment]
The solution for GNU Radio 3.X would be option (3): add another `make_???()` function that includes a channel offset.

Your use case makes perfect sense - it just hasn't come up.
